### ISSUE TITLE
Frequent crashes

### ISSUE THREAD

**Issue by kofa73** [darktable-org/darktable#13893]
**Describe the bug/issue**
Frequesnt crashes during editing. After restart, the same images don't cause a crash.
Backtrace is empty; I'm attaching the log captured from the console.

[dt.log](https://github.com/darktable-org/darktable/files/10948583/dt.log)
The referenced log file, /tmp/darktable_bt_5J9901.txt, is empty, save for:

> this is darktable 4.3.0+1130~gf50cf5ba68 reporting a segfault:

**To Reproduce**
Editing images, switching between images.

With the files in:
https://tech.kovacs-telekes.org/files/darktable-crash-13893/
Switching between the files with space, without editing, triggers the crash. Sometimes I can get through the whole album, sometimes it crashes on a random image. I try to switch between the images as quickly as possible, always going to the next.

I could try a bisect, or provide more info if the attached log does not provide sufficient info. **Bisect pending.**

Self-compiled, Ubuntu 22.10, 'Release' mode. Revision f50cf5ba68. OpenCL on/off does not matter.

**Comment by Nilvus** [comment]
No crash on my side. I don't have Nvidia card. Did you try without OpenCL? And/or with a different Nvidia driver. Maybe it's not OpenCL but we've seen crashes related previously to OpenCL or bad driver version.

**Comment by pehar1** [comment]
Same here, using 4.3.0-1143-g6e46d025b4, no crashes during more than one hour of editing differnt images. Ubuntu 22.04 and Nvidia driver 470.161.03. I decided to install the older driver last august because of the reports of difficulties with version 525. I do not know if all the problems encountered with the 525 driver at that time are fixed.

**Comment by Nilvus** [comment]
So maybe issue from 525 Nvidia driver. @kofa73, try to downgrade (or upgrade if there's more recent release) your driver.

**Comment by pehar1** [comment]
Have you tried with an empty config ? Just to make sure that database and darktablerc are not corrupted. How long have you been using the 525 driver ? Have you had any problems with it so far?

**Comment by jenshannoschwalm** [comment]
I wouldn't bet on the Nvidia driver, 525 has been stable now for a long time and the issues have been resolved too.

**Comment by kofa73** [comment]
Sorry, I was mostly away for the weekend. I'll try to test this properly. Until then, I'm closing it, and will reopen if rolling back the NVidia driver does not fix it. I'll try to bisect in that case.

**Comment by kofa73** [comment]
I've managed to reproduce this without OpenCL: `./darktable --disable-opencl`.

**Comment by pehar1** [comment]
Fresh install of 4.3.0+1143~g6e46d025b4 (empty `~/.config/darktable`), OPENCL active
With the images (**including** the xmps) you referenced in OP I get : 

- error messages for every image regarding missing `.cube` an/or `.png` files (LUT3D)
- can reproduce crashes by (fast) switching between images by selecting them in filmstrip
- could reproduce a crash by using spacebar (more difficult)

With the images **excluding** the xmps (i.e. default history with minimal set of modules) or after lighttable -> discard history for all images (i.e. duplicates still present) I can **not** reproduce any crashes even not by clicking in filmstrip or hitting spacebar extremely fast.

Removing LUT3D from history stack removes the error messages but does not prevent the crashes.

**Comment by kofa73** [comment]
@pehar1 : Thanks for the check. The LUT3D is just a left-over (I was playing with LUTs, but disabled them in the end, so they should not affect the processing).

**Comment by TurboGit** [comment]
@kofa73 @pehar1 : Can you check #13911? Not sure it will help or not...

**Comment by kofa73** [comment]
Hi, #13911 did not prevent the crash.

**Comment by TurboGit** [comment]
I had no big hope on this one :) I'll merge anyway... Feel free to add another log based on master version. Maybe there is multiple issues.

**Comment by kofa73** [comment]
Now this is getting weird. I can reproduce it with an old version (4720525c40), but only if I have my original config in place, not with a clean ~/.config/darktable. Yet, @pehar1 can reproduce it even then. I'll also try with master and a clean config.

**Comment by kofa73** [comment]
```
f4356b88086fb483275d1631b2443f43222e632f is the first bad commit
commit f4356b88086fb483275d1631b2443f43222e632f
Author: Victor Forsiuk <vvforce@gmail.com>
Date:   Wed Feb 8 19:06:23 2023 +0200

    Catch a stray line and put it in the right place

 src/common/darktable.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)
```

**Comment by TurboGit** [comment]
Both commits eaa003cd24 & https://github.com/darktable-org/darktable/commit/f4356b88086fb483275d1631b2443f43222e632f are really clean and impossible to be responsible for the crash to me.

**Comment by victoryforce** [comment]
> Bisect -- but it doesn't seem to make sense, at least not to my Java eyes.

Yes, it really doesn't make sense. This commit simply reorders printf calls to reorder lines in response to the --help command line option. It **cannot** cause a crash. So, the crash reproducibility test was not 100% reliable.

**Comment by TurboGit** [comment]
> Yes. I'll have to try again. It's not easy to do bisect on not-quite deterministic crashes. Sometimes I get a crash on the 3rd space, sometimes on the 20th; but the lack of a crash in any number of pressing the space key does not _prove_ that the bug is not there. :-(

I understand, those bugs are really hard to track down!

**Comment by TurboGit** [comment]
@kofa73:

Given the log, can you try hard to reproduce the issue with this patch:

```
diff --git a/src/develop/imageop.c b/src/develop/imageop.c
index c414a5dc82..4a0a8afb4c 100644
--- a/src/develop/imageop.c
+++ b/src/develop/imageop.c
@@ -1857,7 +1857,7 @@ gboolean _iop_validate_params(dt_introspection_field_t *field,
 static gboolean _iop_update_label(gpointer data)
 {
   dt_iop_module_t *module = (dt_iop_module_t *)data;
-
+#if 0
   const gboolean is_default_params =
     memcmp(module->params, module->default_params, module->params_size) == 0;
 
@@ -1880,6 +1880,7 @@ static gboolean _iop_update_label(gpointer data)
 
   dt_iop_gui_update_header(module);
 
+#endif
   module->label_recompute_handle = 0;
   return G_SOURCE_REMOVE;
 }

```

TIA.

**Comment by kofa73** [comment]
I did press it like crazy... I'll try later.

**Comment by kofa73** [comment]
> Given the log, can you try hard to reproduce the issue with this patch:

Yes, I can :-(

**Comment by TurboGit** [comment]
> Yes, I can :-(

That's somewhat a good news :) I fear that it was due to a recent change of mine :)

So now I'm wondering if this issue is not there since long time actually. I'll try to see if I have another idea to check.

**Comment by pehar1** [comment]
If it is of use I could try tomorrow to reproduce the crashes with dt under gdb. I only have experience with the java debugger but a backtrace should hopefully work.

**Comment by TurboGit** [comment]
@kofa73 : Can you check with this PR #13918. I have a bit more of hope about this one :) TIA.

**Comment by kofa73** [comment]
I get a crash even with 0d7d6773, which is a 3.9 revision, so your suspicion regarding this being a long-standing issue may be right.
(I have to set highlight reconstruction to _reconstruct color_ (maybe it also works with clipping), otherwise some images come up black.)
I'll try your PR.

**Comment by kofa73** [comment]
Nope. :-( 4018189 still crashes.

**Comment by TurboGit** [comment]
Too bad :( @pehar1 yes a gdb session with my PR (https://github.com/darktable-org/darktable/commit/40181890a74b1b18ae18a7ef26db37864a8a60cd) would be nice.

**Comment by kofa73** [comment]
The PR has a weird side-effect: if I disable OpenCL (slow down processing, I guess), keyboard navigation breaks if I try to press space early (before 'working' disappears). Once it's broken, not even the mouse works on the film strip at the bottom. I have to exit to the lighttable and return to the darkroom in order to restore it.

**Comment by pehar1** [comment]
4.3.0-1202-gdd6cb52e9e + patch from pull 13918, OpenCL active, fresh install
Same as @kofa73 reported, after fast pressing of spacebar 5 to 10 times keyboard navigation breaks. Therefore can not force the crashes. Will revert the patch and see if I can get a backtrace after crashing current master. Or should I try something else ?

**Comment by pehar1** [comment]
4.3.0-1202-gdd6cb52e9e, OpenCL active, fresh install
[crash-13893.log](https://github.com/darktable-org/darktable/files/10965068/crash-13893.log)
next working on backtrace...

**Comment by pehar1** [comment]
4.3.0-1202-gdd6cb52e9e, OpenCL active, fresh install
[gdb-backtrace-13893.txt](https://github.com/darktable-org/darktable/files/10965395/gdb-backtrace-13893.txt)
gdb session is still running so I could provide more information if i am guided accordingly

**Comment by TurboGit** [comment]
Please try:

```
(gdb) f 1
(gdb) print *module
```

**Comment by pehar1** [comment]
```
(gdb) f 1
#1  0x00007ffff7be1bfb in _iop_update_label (data=0x55555a27cb50)
    at /home/ph/darktable/master/src/develop/imageop.c:1862
1862        memcmp(module->params, module->default_params, module->params_size) == 0;
(gdb) print *module
$1 = {actions = DT_ACTION_TYPE_CATEGORY, version = 0x0, init_presets = 0x0, 
  init_global = 0x7ffff67fbaf0 <g_cclosure_marshal_VOID__VOIDv>, cleanup_global = 0x1, 
  name = 0x7ffff6800120 <g_cclosure_marshal_VOID__VOID>, aliases = 0x55555bb05a00, 
  default_group = 0x0, flags = 0x7ffff7be85f0 <_display_mask_indicator_callback>, 
  deprecated_msg = 0x41, description = 0x2, operation_tags = 0x555558c04b80, 
  operation_tags_filter = 0x55555561a688, input_format = 0x0, output_format = 0x100000080, 
  default_colorspace = 0x0, input_colorspace = 0x3f3000003be, output_colorspace = 0x91, 
  blend_colorspace = 0x9093a72656b616d, tiling_callback = 0x696e6f73616e6150, 
  gui_update = 0x3a6c65646f6d0a63, gui_reset = 0x584c2d434d440909, 
  gui_init = 0xa29333a34282037, color_picker_apply = 0x9093a746e756f6d, 
    gui_changed = 0x696e6f73616e6170, gui_cleanup = 0x6f72630a37584c63, gui_post_expose = 0x726f746361662070,                                                                                     
  gui_focus = 0x312c35093a, mouse_actions = 0x55555ca26c80, 
  mouse_leave = 0x7ffff79aace0 <main_arena+96>, mouse_moved = 0x55555a27cc88, 
    button_released = 0x0, button_pressed = 0x0, scrolled = 0x0, configure = 0x1, init = 0x91, 
  cleanup = 0x7265626d756e2d66, init_pipe = 0x7574726570612820, commit_params = 0x7fff00296572, 
  reload_defaults = 0x555558e1b1e0, change_image = 0x555558e1b870, 
    cleanup_pipe = 0x555558e1bba0, modify_roi_in = 0x555558e1bd50, modify_roi_out = 0x555558e1cbf0,                                                                                               
  legacy_params = 0x555558e1cf20, masks_selection_changed = 0x0, process = 0x0, 
  process_tiling = 0x0, process_cl = 0x1, process_tiling_cl = 0x2c30ffff0000001f, 
  distort_transform = 0x65636e6172626976, distort_backtransform = 0x0, distort_mask = 0x90, 
  introspection_init = 0x91, get_introspection = 0x2073726564696c73, 
  get_introspection_linear = 0x6565726720726f66, get_p = 0x656e6e616863206e, get_f = 0x6c, 
  set_preferences = 0x0, module = 0x0, op = '\000' <wiederholt 16 Mal>, "\001\000\000", 
  instance = 0, iop_order = 0, hide_enable_button = 0, 
  request_color_pick = (unknown: 0x5ccd44a0), 
  request_histogram = (DT_REQUEST_ON | unknown: 0x5554), request_mask_display = 3, 
  suppress_mask = 0, picked_color = {4.20389539e-45, 0, 0, 0}, picked_color_min = {0, 0, 
    1.85266005e+28, 8.98909342e+14}, picked_color_max = {2.01786979e-43, 0, 2.03188277e-43, 0}, 
  picked_output_color = {2.62214876e+20, 6.96922855e+22, 4.61666145e+24, 4.58727847e+27}, 
  picked_output_color_min = {0.000900270941, 1.73809754e+25, 1.93459244e-19, 1.88876912e+31}, 
  picked_output_color_max = {4.41612452e+21, 3.17751191e+30, 6.83064243e+22, 2.55846603e-12}, 
  histogram = 0x6d6d372c37312d37, histogram_stats = {bins_count = 1701863690, 
--Type <RET> for more, q to quit, c to continue without paging--
```

**Comment by TurboGit** [comment]
You need to press RET and copy/paste the remaining. Some data I'm interested in are not yet displayed.

**Comment by pehar1** [comment]
Sorry, here is the complete output of print...
```
(gdb) print *module
$1 = {actions = DT_ACTION_TYPE_CATEGORY, version = 0x0, init_presets = 0x0, 
  init_global = 0x7ffff67fbaf0 <g_cclosure_marshal_VOID__VOIDv>, cleanup_global = 0x1, 
  name = 0x7ffff6800120 <g_cclosure_marshal_VOID__VOID>, aliases = 0x55555bb05a00, 
  default_group = 0x0, flags = 0x7ffff7be85f0 <_display_mask_indicator_callback>, 
  deprecated_msg = 0x41, description = 0x2, operation_tags = 0x555558c04b80, 
  operation_tags_filter = 0x55555561a688, input_format = 0x0, output_format = 0x100000080, 
  default_colorspace = 0x0, input_colorspace = 0x3f3000003be, output_colorspace = 0x91, 
  blend_colorspace = 0x9093a72656b616d, tiling_callback = 0x696e6f73616e6150, 
  gui_update = 0x3a6c65646f6d0a63, gui_reset = 0x584c2d434d440909, 
  gui_init = 0xa29333a34282037, color_picker_apply = 0x9093a746e756f6d, 
    gui_changed = 0x696e6f73616e6170, gui_cleanup = 0x6f72630a37584c63, gui_post_expose = 0x726f746361662070,                                                                                     
  gui_focus = 0x312c35093a, mouse_actions = 0x55555ca26c80, 
  mouse_leave = 0x7ffff79aace0 <main_arena+96>, mouse_moved = 0x55555a27cc88, 
    button_released = 0x0, button_pressed = 0x0, scrolled = 0x0, configure = 0x1, init = 0x91, 
  cleanup = 0x7265626d756e2d66, init_pipe = 0x7574726570612820, commit_params = 0x7fff00296572, 
  reload_defaults = 0x555558e1b1e0, change_image = 0x555558e1b870, 
    cleanup_pipe = 0x555558e1bba0, modify_roi_in = 0x555558e1bd50, modify_roi_out = 0x555558e1cbf0,                                                                                               
  legacy_params = 0x555558e1cf20, masks_selection_changed = 0x0, process = 0x0, 
  process_tiling = 0x0, process_cl = 0x1, process_tiling_cl = 0x2c30ffff0000001f, 
  distort_transform = 0x65636e6172626976, distort_backtransform = 0x0, distort_mask = 0x90, 
  introspection_init = 0x91, get_introspection = 0x2073726564696c73, 
  get_introspection_linear = 0x6565726720726f66, get_p = 0x656e6e616863206e, get_f = 0x6c, 
  set_preferences = 0x0, module = 0x0, op = '\000' <wiederholt 16 Mal>, "\001\000\000", 
  instance = 0, iop_order = 0, hide_enable_button = 0, 
  request_color_pick = (unknown: 0x5ccd44a0), 
  request_histogram = (DT_REQUEST_ON | unknown: 0x5554), request_mask_display = 3, 
  suppress_mask = 0, picked_color = {4.20389539e-45, 0, 0, 0}, picked_color_min = {0, 0, 
    1.85266005e+28, 8.98909342e+14}, picked_color_max = {2.01786979e-43, 0, 2.03188277e-43, 0}, 
  picked_output_color = {2.62214876e+20, 6.96922855e+22, 4.61666145e+24, 4.58727847e+27}, 
  picked_output_color_min = {0.000900270941, 1.73809754e+25, 1.93459244e-19, 1.88876912e+31}, 
  picked_output_color_max = {4.41612452e+21, 3.17751191e+30, 6.83064243e+22, 2.55846603e-12}, 
  histogram = 0x6d6d372c37312d37, histogram_stats = {bins_count = 1701863690, 
--Type <RET> for more, q to quit, c to continue without paging--
    buf_size = 4047949004481837669, pixels = 1869767434, ch = 1634082928}, histogram_max = {
    1919906915, 741673274, 2037647921, 154822000}, histogram_cst = 1667584521, 
  histogram_middle_grey = 1768712564, dev = 0x756f6d0a7261656e, enabled = 980644974, 
  default_enabled = 1851879433, params = 0x584c63696e6f7361, default_params = 0x37, 
  params_size = 81, gui_data = 0x0, gui_lock = {mutex = {__data = {__lock = 0, __count = 0, 
        __owner = 0, __nusers = 0, __kind = -159373360, __spins = 32767, __elision = 0, 
        __list = {__prev = 0x1, __next = 0x7ffff6800900 <g_cclosure_marshal_VOID__BOXED>}}, 
      __size = '\000' <wiederholt 16 Mal>, "\320'\200\366\377\177\000\000\001\000\000\000\000\000\000\000\000\t\200\366\377\177\000", __align = 0}}, global_data = 0x55555bb49c90, 
  blend_params = 0x0, default_blendop_params = 0x7ffff6e388a0, blend_data = 0x41, 
  raster_mask = {source = {Python Exception <class 'gdb.error'>: Es gibt kein Mitglied mit dem Namen keys.
users = 0x2, Python Exception <class 'gdb.error'>: Es gibt kein Mitglied mit dem Namen keys.
masks = 0x555558b4d0c0}, sink = {
      source = 0x555555619c08, id = 1488962182}}, widget = 0x100000080, off = 0x0, 
  header = 0x926, label = 0x31, instance_name = 0x0, mask_indicator = 0x0, 
  expander = 0xb0d4b882d03f5681, expanded = 0, reset_button = 0x0, presets_button = 0x41, 
  fusion_slider = 0x2, widget_list = 0x0, widget_list_bh = 0x55555561a935 = {
    0x6172006b63656863, 
    <Fehler beim Lesen der Variable: Cannot access memory at address 0x626174006f6964>
```

**Comment by TurboGit** [comment]
This crash is very different than the previous one. I have added a check in the PR. Please check again when you have time. TIA.

**Comment by TurboGit** [comment]
@pehar1 : Don't spent time on my PR, it is wrong... I'll need more thinking.

**Comment by pehar1** [comment]
Take all the time you need, I will probably not be able to test further until tomorrow.

**Comment by TurboGit** [comment]
Good news, I have been able to reproduce a crash when having lot of multiple instances on a module. I think I understand what is going on and the new PR fixes the issue on my side. So if we had the same issue this is probably the right & finale PR.

**Comment by kofa73** [comment]
I'll try to test this.

**Comment by kofa73** [comment]
I find it easier to reproduce with OpenCL, profile set to 'very fast GPU', optimise for 'memory', and the filmstrip hidden.
It took a while to reproduce this with gdb, but then I was unable to switch between windows (I was able to move the mouse pointer, and also to Ctrl+Alt+F2 to a new login console).
Maybe I should run the gdb session in 'screen', so I can then pull the screen session over to the character console? Any other ideas?

**Comment by TurboGit** [comment]
> I'll try to test this.
> 
> Update: :-(

Are you sure you have the right version? Above I ask to test a code deactivating some part and you were able to reproduce the issue (https://github.com/darktable-org/darktable/issues/13893#issuecomment-1465896347), but clearly (since I was able to reproduce) I tested my patch above and the issue was really in a part that should have been deactivated. So either there is some mis-compilation or you have a separate issue.

Let's wait for @pehar1 check too.

**Comment by kofa73** [comment]
I'm pretty sure it was a clean build of your PR (check the rev-parse; I made sure to have a clean workspace before building). However, I've also had the feeling there might be two separate issues. I'll try downgrading the Nvidia driver tomorrow.

**Comment by kofa73** [comment]
I'll try to check again tomorrow, and maybe also downgrade the Nvidia driver.

**Comment by kofa73** [comment]
Darktable reports 4.3.0+1204~g14f43bc811.
I'm pretty sure it was a clean build:

    kofa@eagle:~/darktable$ git status
    On branch po/fix-fast-img-reload
    nothing to commit, working tree clean
    kofa@eagle:~/darktable$ git rev-parse HEAD
    14f43bc811e89459a5192b58ab3d8cd266cefee4

I'll check the driver.

**Comment by pehar1** [comment]
4.3.0+1202~gdd6cb52e9e (same as yesterday) + patches from PR13920 in `imageop.c` (manually added)

No crashes anymore no matter how crazy I peck at the keys, also stable when changing selections very quickly via the filmstrip. For me the issue seems solved. :-))
 Thank you @TurboGit !

> lot of multiple instances on a module

This explains why the problem could not be reproduced after a history reset. Also not on my own images which also have only "simpler" histories.

**Comment by TurboGit** [comment]
@pehar1 : Great news, the fix is merged now.

FYI, I'm using NVidia driver version 525.89.02 on Debian without issue.

**Comment by kofa73** [comment]
I had 525.85.05 on Ubuntu 22.10. Maybe more than the compilation issue mentioned in https://discuss.pixls.us/t/fixed-gnu-debian-nvidia-opencl-520-x-525-x/35321 was fixed in 525.89.
